### PR TITLE
[ntuple] Move RColumnElement concrete definitions out of the header file

### PR DESCRIPTION
**Pull request by silverweed**
# This Pull request:
by @hahnjo suggestion, since we are exposing a type-erased virtual interface for `RColumnElement` through `RColumnElementBase`, we can hide all the concrete definitions into a cxx file.
This has the benefit of reducing the compile time of any translation unit that includes `RColumnElementBase.hxx` significantly, as the compiler won't have to instantiate all the combination matrix of `RColumnElement<CppT, ColumnType>` every time.
It also let us have better control on exactly which types of `RColumnElement` we allow to instantiate by explicitly listing them into a proxy enum.

## Changes or fixes:
- renamed `RColumnElement.hxx` to `RColumnElementBase.hxx`
- moved all concrete definitions from `RColumnElementBase.hxx` to `src/RColumnElement.hxx`. This is a private header file that is included by `RColumnElement.cxx` and by `ntuple_endian.cxx`. The reason to separate this from `RColumnElement.cxx` is that `ntuple_endian.cxx` needs to simulate a big-endian machine by defining `R__LITTLE_ENDIAN 0`, which changes the definitions of some `RColumnElement`s. To avoid including the whole `RColumnElement.cxx` in the test, we decided to split the definitions into a file that can be included independently by the test. It's not a perfect solution (the test executable ends up with mismatching instantiations of RColumnElement since it links to libROOTNTuple.so) but it's technically not worse than before. We still might want to think of alternative solutions.
- introduces an enum `EColumnCppType` that lists all the allowed c++ in-memory types for RColumnElement. This is used internally to map the templated `RColumnElementBase::Generate` to a non-templated function that can be implemented in the cxx file.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RColumnElementBase.hxx [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+   else if constexpr (std::is_same_v<CppT, std::int32_t>)
+      return GenerateColumnElement(EColumnCppType::kInt32, type);
+   else if constexpr (std::is_same_v<CppT, std::int64_t>)
+      return GenerateColumnElement(EColumnCppType::kInt64, type);
+   else if constexpr (std::is_same_v<CppT, float>)
+      return GenerateColumnElement(EColumnCppType::kFloat, type);
+   else if constexpr (std::is_same_v<CppT, double>)
+      return GenerateColumnElement(EColumnCppType::kDouble, type);
+   else if constexpr (std::is_same_v<CppT, Double32_t>)
+      return GenerateColumnElement(EColumnCppType::kDouble32, type);
```
Comment (hahnjo) [inline tree/ntuple/v7/inc/ROOT/RColumnElementBase.hxx:155]: `Double32_t` is a `typedef`, so for what concerns C++ it is a `double` and this check should never be true.

File: tree/ntuple/v7/inc/ROOT/RColumnElementBase.hxx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+   else if constexpr (std::is_same_v<CppT, ClusterSize_t>)
+      return GenerateColumnElement(EColumnCppType::kClusterSize, type);
+   else if constexpr (std::is_same_v<CppT, RColumnSwitch>)
+      return GenerateColumnElement(EColumnCppType::kColumnSwitch, type);
+   else if constexpr (std::is_same_v<CppT, void>)
+      return Generate<void>(type);
```
Comment (hahnjo) [inline tree/ntuple/v7/inc/ROOT/RColumnElementBase.hxx:161]: This should never happen because of the explicit specialization.

File: tree/ntuple/v7/src/RColumnElement.hxx [hunk 3]
```diff
@@ -240,102 +222,80 @@ static void CastZigzagSplitUnpack(void *destination, const void *source, std::si
...
+} // namespace
 
-} // anonymous namespace
+// anonymous namespace because these definitions are not meant to be exported.
+namespace {
 
-namespace ROOT {
-namespace Experimental {
-namespace Internal {
+using ROOT::Experimental::EColumnType;
```
Comment (hahnjo) [inline tree/ntuple/v7/src/RColumnElement.hxx:230]: You may consider also `using ROOT::Experimental::Internal::RColumnElementBase` to avoid some churn

**Review by hahnjo — COMMENTED** [review]
I'm convinced this is a good direction, but I'm probably biased here :wink: 

For the commit organization, it's not ideal that the first commit changes `ntuple_endian.cxx` to `RColumnElementBase::Generate` (which we know will fail the test) and the second commit reverts the change. It might also be good to update the message and be clear that the definitions are not actually moved in that change (yet)

File: tree/ntuple/v7/src/RColumnElement.hxx [hunk 4]
```diff
@@ -1,38 +1,20 @@
...
+// RColumnElement concrete implementations
+//
+// Note that this file is in the src directory and not in the inc directory because we need the ability
+// to override R__LITTLE_ENDIAN for testing purposes.
+// This is not a particularly clean or correct solution, as the tests that do this will end up with two different
+// definitions of some RColumnElements, so we might want to change this mechanism in the future. In any case, these
+// definitions are implementation details and should not be exposed to a public interface.
+
+#include <ROOT/RColumnElementBase.hxx>
+#include <ROOT/EExecutionPolicy.hxx>
```
Comment (jblomer) [inline tree/ntuple/v7/src/RColumnElement.hxx:10]: Why do we need this include?

**Review by jblomer — COMMENTED** [review]
The approach looks good to me.

File: tree/ntuple/v7/src/RColumnElement.hxx (hunk 4 above)
Comment (silverweed) [reply tree/ntuple/v7/src/RColumnElement.hxx:10]: Not sure, it was added automatically by my editor. That's an annoying functionality that I need to disable because it's so easy to miss such things.

File: tree/ntuple/v7/inc/ROOT/RColumnElementBase.hxx [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+   virtual bool IsMappable() const
+   {
+      R__ASSERT(false);
+      return false;
+   }
+
+   /// If the on-storage layout and the in-memory layout differ, packing creates an on-disk page from an in-memory page
+   virtual void Pack(void *destination, const void *source, std::size_t count) const
+   {
+      std::memcpy(destination, source, count);
```
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RColumnElementBase.hxx:89]: These warning predate the PR. Still, can we do something about them? Can be a separate PR.

**Review by jblomer — APPROVED** [review]
LGTM, but see comment about warnings.

File: tree/ntuple/v7/inc/ROOT/RColumnElementBase.hxx (hunk 5 above)
Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RColumnElementBase.hxx:89]: I'll try to address them in a future PR, thanks for pointing them out.
